File: 0006-zigzag-conversion/0006-zigzag-conversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if(numRows == 1) return s; 
        vector<string> rows(numRows); 
        int currRow = 0; 
        bool flag = true; 
        for(int i = 0; i < s.length(); i++){
            rows[currRow] += s[i]; 

            if(currRow >= numRows - 1){
                flag = false; 
            }
            if(currRow == 0){
                flag = true; 
            }

            if(flag){
                currRow += 1; 
            } else{
                currRow -= 1; 
            }
        }

        string answer = ""; 
        for(string next : rows) answer += next; 
        return answer; 
    }
};

File: 0006-zigzag-conversion/0006-zigzag-conversion.java
class Solution {
    public String convert(String s, int numRows) {
        if(numRows == 1) return s; 
        StringBuilder[] sbArray = new StringBuilder[numRows]; 
        for(int i = 0; i < numRows; i++){
            sbArray[i] = new StringBuilder(); 
        }

        int curr = 0; 
        int dir = 1; 
        for(char c : s.toCharArray()){
            sbArray[curr].append(c); 
            
            if(dir == 1){
                if(curr + 1 >= numRows){
                    curr--; 
                    dir = -1; 
                } else{
                    curr++; 
                }
            } else if(dir == -1){
                if(curr - 1 < 0){
                    curr++; 
                    dir = 1; 
                } else{
                    curr--; 
                }
            }
        }
        //System.out.print(Arrays.toString(sbArray)); 
        return String.join("", sbArray); 
    }
}